File: python/src/bedrock_bio/load_table.py
import duckdb

from .config import config


def load_table(name: str) -> duckdb.DuckDBPyRelation:
    """Lazily query a table.

    Parameters
    ----------
    name : str
        Table identifier.

    Returns
    -------
    duckdb.DuckDBPyRelation
        A lazy relation that can be further filtered, selected, or collected.
        Filter on partition columns (see ``describe_table()``) for fastest reads.

    Examples
    --------
    >>> import bedrock_bio as bb
    >>> df = (
    ...     bb.load_table("dbsnp.vcf")
    ...     .filter("assembly = 'GRCh38' AND chromosome = '22'")
    ...     .select("rsid, position, ref_allele, alt_allele")
    ...     .limit(5)
    ...     .df()
    ... )
    """
    manifest = config.get_manifest()
    if name not in manifest:
        raise ValueError(
            f"Table '{name}' not found in manifest. "
            f"See list_tables() for available tables."
        )

    conn = config.get_connection()
    escaped = manifest[name]["iceberg_json"].replace("'", "''")
    return conn.sql(f"SELECT * FROM iceberg_scan('{escaped}')")

File: r/R/load_table.R
#' Lazily query a table
#'
#' @param name Table identifier.
#' @returns A lazy `tbl` backed by DuckDB, compatible with dplyr verbs. Filter
#'   on partition columns (see `describe_table()`) for fastest reads.
#'
#' @examples
#' \dontrun{
#' library(bedrockbio)
#' library(dplyr)
#'
#' load_table("dbsnp.vcf") |>
#'   filter(assembly == "GRCh38", chromosome == "22") |>
#'   select(rsid, position, ref_allele, alt_allele) |>
#'   head(5) |>
#'   collect()
#' }
#'
#' @export
load_table <- function(name) {
  manifest <- get_manifest()
  if (!name %in% names(manifest)) {
    stop(
      "Table '", name, "' not found in manifest. ",
      "See list_tables() for available tables.",
      call. = FALSE
    )
  }

  conn <- get_connection()
  query <- DBI::sqlInterpolate(
    conn,
    "SELECT * FROM iceberg_scan(?path)",
    path = manifest[[name]]$iceberg_json
  )
  dplyr::tbl(conn, dplyr::sql(query))
}
